torchreid/engine/image/multilabel.py

```diff
@@ -0,0 +1,191 @@
+                        kl_loss = 0
+                        for model_j, logits_j in enumerate(out_logits[trg_id]):
+                            if model_i != model_j:
+                                probabilities_j = torch.sigmoid(logits_j)
+                                kl_loss += self.kl_div_binary(probabilities_i, probabilities_j)
+                        mutual_loss += kl_loss / (model_num - 1)
+                        loss_summary['mutual_{}/{}'.format(trg_id, model_names[model_i])] = mutual_loss.item()
+
+                should_turn_off_mutual_learning = self._should_turn_off_mutual_learning(self.epoch)
+                coeff_mutual_learning = int(not should_turn_off_mutual_learning)
+
+                total_loss += coeff_mutual_learning * mutual_loss
+            # backward pass
+            total_loss.backward(retain_graph=False)
+
+            for model_name in model_names:
+                if not isinstance(self.optims[model_name], SAM) and step == 1:
+                    self.optims[model_name].step()
+                elif step == 1:
+                    assert isinstance(self.optims[model_name], SAM)
+                    self.optims[model_name].first_step()
+                else:
+                    assert isinstance(self.optims[model_name], SAM) and step==2
+                    self.optims[model_name].second_step()
+
+            loss_summary['loss'] = total_loss.item()
+
+        return loss_summary, avg_acc
+
+    def _single_model_losses(self, model, train_records, imgs, obj_ids, n_iter, model_name):
+        model_output = model(imgs)
+        all_logits = self._parse_model_output(model_output)
+
+        total_loss = torch.zeros([], dtype=imgs.dtype, device=imgs.device)
+        out_logits = []
+        loss_summary = dict()
+
+        num_trg_losses = 0
+        avg_acc = 0
+
+        for trg_id in range(self.num_targets):
+            trg_mask = train_records['dataset_id'] == trg_id
+
+            trg_obj_ids = obj_ids[trg_mask]
+            trg_num_samples = trg_obj_ids.numel()
+            if trg_num_samples == 0:
+                out_logits.append(None)
+                continue
+
+            trg_logits = all_logits[trg_id][trg_mask]
+            main_loss = self.main_losses[trg_id](trg_logits, trg_obj_ids)
+            avg_acc += metrics.accuracy_multilabel(trg_logits, trg_obj_ids).item()
+            loss_summary['main_{}/{}'.format(trg_id, model_name)] = main_loss.item()
+
+            scaled_trg_logits = self.main_losses[trg_id].get_last_scale() * trg_logits
+            out_logits.append(scaled_trg_logits)
+
+            total_loss += main_loss
+            num_trg_losses += 1
+
+        total_loss /= float(num_trg_losses)
+        avg_acc /= float(num_trg_losses)
+
+        return total_loss, loss_summary, avg_acc, out_logits
+
+    def kl_div_binary(self, x, y):
+        ''' compute KL divergence between two tensors represented
+        independent binary distributions'''
+        y_log = torch.log(y)
```

Comment:
If you pass targets in log space, you have to use the corresponding flag.
Seems like this formula is incorrect: kl_div expects that you have both p and 1-p in the input tensor.

Reply:
If I understand correctly the documentations: https://pytorch.org/docs/1.9.0/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss
KLdivLoss(x,y) where x must be log probs and y must be just probabilities. Maybe I messed up the order. But I think it doesn't matter as we compute KL(x | y) and KL(y | x). I will change the x and y position.